packages/gatsby/cache-dir/production-app.js

```diff
@@ -71,7 +71,13 @@ apiRunnerAsync(`onClientEntry`).then(() => {
       if (e.page.path === loader.getPage(pathname).path) {
         emitter.off(`onPostLoadPageResources`, eventHandler)
         clearTimeout(timeoutId)
-        window.___history.push(pathname)
+
+        // Only when `window.___history` exist, then navigate using it
+        if (window.___history) {
+          window.___history.push(pathname)
+        } else {
+          window.location.href = pathname
```

Comment:
Hi @KyleAMathews , based on my investigation, `gatsby-link` uses `window.___navigateTo` to navigate between pages and `window.navigateTo` relies on `window.___history` to navigate. Apparently, when a resource/page not found in the directory, the host will use `404.html` to serve page not found and `window.___history` does not seem to exist. Hence, using `window.location.href` to navigate between pages.